### PR TITLE
Automated testing speeds up global configuration queries

### PR DESCRIPTION
**Pull request by kakj-go**
#### What type of this PR

/kind feature

#### What this PR does / why we need it:
Since the query automation test configuration is a circular query, too many configurations will slow down the automation test componentization, which will seriously affect users' use

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://terminus-org.app.terminus.io/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb24iOls4ODEsNzcyXSwibWVtYmVyIjpbIjEwMDA1NjAiXX0%3D&id=258176&iterationID=772&pId=0&type=BUG)
- [Erda Cloud Issue Link](https://terminus-org.app.terminus.io/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb24iOls4ODEsNzcyXSwibWVtYmVyIjpbIjEwMDA1NjAiXX0%3D&id=258180&iterationID=772&pId=0&type=BUG)

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |       Optimizing query speed for global configuration       |
| 🇨🇳 中文    |      优化全局配置查询速度        |

### CODE REVIEW

File: modules/dop/services/autotest/global_config.go [hunk 1]
```diff
@@ -334,15 +334,82 @@ func (svc *Service) ListGlobalConfigs(req apistructs.AutoTestGlobalConfigListReq
...
+				var uiConfig apistructs.AutoTestUIConfig
+				if err := json.Unmarshal([]byte(cfg.Value), &uiConfig); err != nil {
+					return nil, fmt.Errorf("failed to unmarshal uiConfig, err: %v", err)
+				}
+				result.UIConfig = &uiConfig
+			}
+		}
 
-	for _, ns := range namespaces.Data {
-		cfg, err := svc.parseGlobalConfigFromCmsNs(ns.Ns)
```
Comment (sfwn) [inline modules/dop/services/autotest/global_config.go:340]: you have to reuse the logic of `parseGlobalConfigFromCmsNs`.

File: modules/dop/services/autotest/global_config.go [hunk 2]
```diff
@@ -334,15 +334,82 @@ func (svc *Service) ListGlobalConfigs(req apistructs.AutoTestGlobalConfigListReq
...
+		config := cfg
+		if config.PipelineCmsNs == nil {
+			continue
+		}
+		nsCfgMap[config.PipelineCmsNs.Ns] = append(nsCfgMap[config.PipelineCmsNs.Ns], config)
+	}
+
 	var sortResult apistructs.SortByUpdateTimeAutoTestGlobalConfigs
+	for ns, configs := range nsCfgMap {
+		result := apistructs.AutoTestGlobalConfig{Ns: ns}
```
Comment (sfwn) [inline modules/dop/services/autotest/global_config.go:363]: too bad to copy these code again.

File: api/proto/core/pipeline/cms/cms.proto [hunk 3]
```diff
@@ -38,13 +38,20 @@ service CmsService {
...
+    };
+  }
 }
 
 message PipelineCmsNs {
   string pipeline_source = 1;
   string ns = 2;
   google.protobuf.Timestamp timeCreated = 3;
   google.protobuf.Timestamp timeUpdated = 4;
+  uint64 ID = 5 [json_name = "id"];
```
Comment (sfwn) [inline api/proto/core/pipeline/cms/cms.proto:54]: no id needed. `ns` is enough.

File: modules/pipeline/providers/cms/cms.service.go [hunk 4]
```diff
@@ -138,6 +138,28 @@ func (s *cmsService) GetCmsNsConfigs(ctx context.Context, req *pb.CmsNsConfigsGe
 		})
 	}
 
+	// order by timeCreated
+	sortResult(results)
```
Comment (sfwn) [inline modules/pipeline/providers/cms/cms.service.go:142]: no need sort at pipeline level, just return with input order.

Or all use two-layer map, source for first-layer and ns for second-layer.

File: modules/pipeline/providers/cms/db/op.go [hunk 5]
```diff
@@ -92,6 +92,17 @@ func (client *Client) GetCmsNs(pipelineSource apistructs.PipelineSource, ns stri
 	return cmsNs, exist, nil
 }
 
+func (client *Client) BatchGetCmsNs(pipelineSource []string, ns []string, ops ...mysqlxorm.SessionOption) ([]PipelineCmsNs, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/cms/db/op.go:95]: use Plural forms of words.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: api/proto/core/pipeline/cms/cms.proto [hunk 6]
```diff
@@ -147,6 +154,17 @@ message CmsNsConfigsGetRequest {
   repeated PipelineCmsConfigKey keys = 3; // only fetch specified keys
   bool globalDecrypt = 4; // global-level setting's priority is lower than key-level
 }
+
+message CmsNsConfigsBatchGetRequest {
+    repeated CmsNsConfigsGetRequest arrays = 1;
```
Comment (sfwn) [inline api/proto/core/pipeline/cms/cms.proto:159]: Can you query cms configs cross ns?

File: modules/pipeline/providers/cms/db/op.go [hunk 7]
```diff
@@ -92,6 +92,17 @@ func (client *Client) GetCmsNs(pipelineSource apistructs.PipelineSource, ns stri
 	return cmsNs, exist, nil
 }
 
+func (client *Client) BatchGetCmsNamespaces(pipelineSource []string, ns []string, ops ...mysqlxorm.SessionOption) ([]PipelineCmsNs, error) {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	var namespaces []PipelineCmsNs
+	if err := session.In("pipeline_source", pipelineSource).In("ns", ns).Find(&namespaces); err != nil {
```
Comment (sfwn) [inline modules/pipeline/providers/cms/db/op.go:100]: You will get many unexpected result.
I think the size of source can only be one.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/cms/db/op.go [hunk 8]
```diff
@@ -92,6 +92,17 @@ func (client *Client) GetCmsNs(pipelineSource apistructs.PipelineSource, ns stri
 	return cmsNs, exist, nil
 }
 
+func (client *Client) BatchGetCmsNamespaces(pipelineSource string, ns []string, ops ...mysqlxorm.SessionOption) ([]PipelineCmsNs, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/cms/db/op.go:95]: namespaces

File: modules/pipeline/providers/cms/pipeline_cms.go [hunk 9]
```diff
@@ -270,6 +270,108 @@ func (c *pipelineCm) GetConfigs(ctx context.Context, ns string, globalDecrypt bo
 	return result, nil
 }
 
+func (c *pipelineCm) BatchGetConfigs(ctx context.Context, req *pb.CmsNsConfigsBatchGetRequest) ([]*pb.PipelineCmsConfig, error) {
+
+	var ns []string
```
Comment (sfwn) [inline modules/pipeline/providers/cms/pipeline_cms.go:275]: namespaces

File: modules/pipeline/providers/cms/pipeline_cms.go [hunk 10]
```diff
@@ -270,6 +270,108 @@ func (c *pipelineCm) GetConfigs(ctx context.Context, ns string, globalDecrypt bo
...
+		}
+
+		newConfig.PipelineCmsNs = &pb.PipelineCmsNs{
+			PipelineSource: relationNs.PipelineSource.String(),
+			Ns:             relationNs.Ns,
+			TimeCreated:    getPbTimestamp(config.TimeCreated),
+			TimeUpdated:    getPbTimestamp(config.TimeUpdated),
+		}
+
+		getReq, ok := sourceKeyMap[relationNs.PipelineSource.String()+"_"+relationNs.Ns]
```
Comment (sfwn) [inline modules/pipeline/providers/cms/pipeline_cms.go:332]: strange map

File: modules/pipeline/providers/cms/pipeline_cms.go [hunk 11]
```diff
@@ -270,6 +270,108 @@ func (c *pipelineCm) GetConfigs(ctx context.Context, ns string, globalDecrypt bo
...
+	}
+	configs, err := c.dbClient.BatchGetCmsNsConfigs(cmsNsIDList, transformKeysToStrSlice(keys...))
+	if err != nil {
+		return nil, err
+	}
+
+	var newConfigs []*pb.PipelineCmsConfig
+	for _, config := range configs {
+
+		var newConfig = pb.PipelineCmsConfig{
```
Comment (sfwn) [inline modules/pipeline/providers/cms/pipeline_cms.go:307]: you should reuse the logic of config decode.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: api/proto/core/pipeline/cms/cms.proto [hunk 12]
```diff
@@ -100,6 +106,7 @@ message PipelineCmsConfig {
   string from = 7;
   google.protobuf.Timestamp timeCreated = 8;
   google.protobuf.Timestamp timeUpdated = 9;
+  PipelineCmsNs pipelineCmsNs = 10;
```
Comment (sfwn) [inline api/proto/core/pipeline/cms/cms.proto:109]: just `ns`.

File: api/proto/core/pipeline/cms/cms.proto [hunk 13]
```diff
@@ -38,6 +38,12 @@ service CmsService {
       get: "/api/pipelines/cms/ns/{ns}",
     };
   }
+
+  rpc BatchGetCmsNsConfigs (CmsNsConfigsBatchGetRequest) returns (CmsNsConfigsBatchGetResponse) {
+    option (google.api.http) = {
+      get: "/api/pipelines/cms/ns/action/batch-get-config",
```
Comment (sfwn) [inline api/proto/core/pipeline/cms/cms.proto:44]: /api/pipelines/cms/actions/batch-get-configs

File: api/proto/core/pipeline/cms/cms.proto [hunk 14]
```diff
@@ -147,6 +154,19 @@ message CmsNsConfigsGetRequest {
...
+    repeated string namespaces = 2;
+    bool globalDecrypt = 3; // global-level setting's priority is lower than key-level
+}
+
 message CmsNsConfigsGetResponse {
   repeated PipelineCmsConfig data = 1;
 }
+
+message CmsNsConfigsBatchGetResponse {
+  repeated PipelineCmsConfig arrays = 1;
```
Comment (sfwn) [inline api/proto/core/pipeline/cms/cms.proto:169]: configs

File: modules/pipeline/providers/cms/db/op.go [hunk 15]
```diff
@@ -170,3 +181,16 @@ func (client *Client) GetCmsNsConfigs(cmsNs PipelineCmsNs, keys []string, ops ..
 	}
 	return configs, nil
 }
+
+func (client *Client) BatchGetCmsNsConfigs(cmsNsID []uint64, ops ...mysqlxorm.SessionOption) ([]PipelineCmsConfig, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/cms/db/op.go:185]: cmsNsIDs

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
And add test screenshot for this.

**Comment by kakj-go** [comment]
> And add test screenshot for this.
CICD:

image

image

image

AUTOTEST:

image


image


image

**Review by sfwn — APPROVED** [review]

**Pull request merged**